src/entity/service.ts

```diff
@@ -0,0 +1,122 @@
+  repository: EntityRepository,
+): Promise<EntityService> {
+  return {
+    getAllEntities: partial(getAllEntities, repository),
+    loadEntities: partial(loadEntities, repository),
+  }
+}
+
+export interface EntityService {
+  getAllEntities: () => Promise<Entity[]>
+  loadEntities: () => Promise<void>
+}
+
+export interface EntityRepository {
+  getAllEntities: () => Promise<Entity[]>
+  findById: (id: string) => Promise<Entity | null>
+  addEntity: (entity: Entity) => Promise<Entity>
+  updateEntity: (entity: Entity) => Promise<Entity>
+}
+
+export type Entity = z.infer<typeof Entity>
+export type EntityDto = z.infer<typeof EntityDto>
+
+export const EntityDto = z
+  .object({
+    id: z.string().openapi({ example: '8fa665b6-7fc5-4b0b-baee-6221b1844ec8' }),
+    name: z.string().openapi({ example: 'Absa' }),
+    did: z.string().openapi({ example: 'did:sov:2NPnMDv5Lh57gVZ3p3SYu3' }),
+    logo_url: z.string().openapi({
+      example:
+        'https://s3.eu-central-1.amazonaws.com/builds.eth.company/absa.svg',
+    }),
+    domain: z.string().openapi({
+      example: 'www.absa.africa',
+    }),
+    role: z
+      .array(z.enum(['issuer', 'verifier']))
+      .openapi({ example: ['issuer', 'verifier'] }),
+    credentials: z
+      .array(z.string())
+      .openapi({ example: ['2NPnMDv5Lh57gVZ3p3SYu3:3:CL:152537:tag1'] }),
+  })
+  .openapi('EntityRequest')
+
+export const Entity = EntityDto.extend({
+  createdAt: z.string().datetime().openapi({ example: '2023-05-24T18:14:24' }),
+  updatedAt: z.string().datetime().openapi({ example: '2023-05-24T18:14:24' }),
+}).openapi('EntityResponse')
+
+async function getAllEntities(repository: EntityRepository) {
+  return repository.getAllEntities()
+}
+
+async function loadEntities(repository: EntityRepository) {
+  const registryContent = await fs.readFile('./src/data/registry.json', {
+    encoding: 'utf8',
+  })
+  const registry = JSON.parse(registryContent)
+  const results = await Promise.allSettled(
+    registry.entities.map(async (e: EntityDto) => {
+      logger.info('Importing entity', e)
+      if (await exists(repository, e)) {
+        logger.debug('Entity already exists, updating...')
+        return updateEntity(repository, e)
+      } else {
+        logger.debug('Entity does not exist, creating...')
+        return addEntity(repository, e)
+      }
+    }),
+  )
+  results.forEach((r) => {
+    if (r.status !== 'fulfilled') {
+      logger.info('Entity import failed with the following error', r.reason)
+    }
+  })
+}
+
+async function exists(repository: EntityRepository, entity: EntityDto) {
+  return repository.findById(entity.id)
+}
+
+async function addEntity(
+  repository: EntityRepository,
+  payload: Record<string, unknown>,
+): Promise<void> {
+  const entityDto = EntityDto.parse(payload)
+  const entity = {
+    ...entityDto,
+    createdAt: new Date().toISOString(),
+    updatedAt: new Date().toISOString(),
+  }
+  await repository.addEntity(entity)
```

Comment:
What happens if we try to add an Entity that already exists in the db?

Reply:
Should we not do a check like we do for `updateEntity`, to ensure that `findById` doesn't exist in this case?

Reply:
> What happens if we try to add an Entity that already exists in the db?

It will throw an error that rejects the promise and then we log it with:

```
results.forEach((r) => {
    if (r.status !== 'fulfilled') {
      logger.info('Entity import failed with the following error', r.reason)
    }
  })
```

Though, I'm not 100 % sure right now if the mapped promises using `async/await` inside `allSettled` behaves exactly like that.

> Should we not do a check like we do for `updateEntity`, to ensure that `findById` doesn't exist in this case?

We already check that with `if (await exists(repository, e)) {`. The check with `existingEntity` is there because I'm retrieving it to construct an updated entity as a whole. I'll look at it. I might not need to do that.

Reply:
Yes, I confirmed that we handle the error as I expected. I also improved the handling a bit. We didn't have any information on what particular entity failed. I added at least an index from the input array for now.